### PR TITLE
Preserve source ip in TPROXY mode

### PR DESCRIPTION
**Pull request by gmemcc**
This PR will:
1. Enable listener filter `original_src` for sidecar inbound listeners, so local process can see real client ip
2. Add iptables rule to ensure correct routing.

Related issue: https://github.com/istio/istio/issues/5679

### CODE REVIEW

**Comment by gmemcc** [comment]
@googlebot I signed it!

File: tools/istio-iptables/pkg/cmd/root.go [hunk 1]
```diff
@@ -44,6 +44,9 @@ var (
 	// istio-proxy to send.
 	dnsVar = env.RegisterStringVar("DNS_CAPTURE", "",
 		"If set, enable the capture of outgoing DNS packets on port 53, redirecting to :15013")
+
+	tproxyPreserveSourceIP = env.RegisterBoolVar("TPROXY_PRESERVE_SOURCE_IP", true,
```
Comment (rlenglet) [inline tools/istio-iptables/pkg/cmd/root.go:48]: Remove this flag. I would do this unconditionally for TPROXY.

File: tools/istio-iptables/pkg/cmd/run.go [hunk 2]
```diff
@@ -427,7 +427,12 @@ func (iptConfigurator *IptablesConfigurator) run() {
 			"-p", "udp", "--dport", "53",
 			"-j", "REDIRECT", "--to-port", "15013")
 	}
-
+	if iptConfigurator.cfg.InboundInterceptionMode == constants.TPROXY && tproxyPreserveSourceIP.Get() {
```
Comment (rlenglet) [inline tools/istio-iptables/pkg/cmd/run.go:430]: Do this unconditionally.

**Review by rlenglet — CHANGES_REQUESTED** [review]
This WIP is looking good so far. :)

Some unit tests will need to be fixed.
And this PR is still missing configuring the `original_src` listener filter.

File: tools/istio-iptables/pkg/cmd/run.go (hunk 2 above)
Comment (gmemcc) [reply tools/istio-iptables/pkg/cmd/run.go:430]: Thanks for the review, @rlenglet
I agree with this suggestion and will take a modification tomorrow. by the way, configuration for original_src also committed, PTAL.

**Comment by costinm** [comment]
Do we have adequate integration tests for TPROXY ? I was planning to propose dropping support for TPROXY and cleaning up the code, this has been around for a long time and it is still in 'pre-alpha' 
mode as testing AFAIK. 

So if you want to add features to tproxy  - first step would be to make sure it doesn't get removed...

**Comment by duderino** [comment]
@costinm what does @gmemcc have to do to ensure TPROXY doesn't get removed?   Just add integration tests?

**Comment by costinm** [comment]
Yes, this has been around for a very long time - integration tests at
least, and some plan to move it forward.

**Comment by gmemcc** [comment]
@rlenglet  Hi, I have tested this WIP with istio 1.5 and it works fine. 

TPROXY itself in istio 1.6 seems doesn't work properly,  I creating an issue for it: https://github.com/istio/istio/issues/23369 . replacing 15001 with 15006 in TPROXY rule can solve the problem:
```bash
iptables -t mangle -R  ISTIO_TPROXY 1 ! -d 127.0.0.1/32 \
    -p tcp -j TPROXY --on-port 15006 --on-ip 0.0.0.0 --tproxy-mark 0x539/0xffffffff
```

File: manifests/charts/global.yaml [hunk 3]
```diff
@@ -57,6 +57,10 @@ global:
   # One central istiod controls all remote clusters: disabled by default
   centralIstiod: false
 
+  # If true, listener filter original_src will be added for inbound listeners of sidecars
+  # running in TPROXY interception mode
+  tproxyPreserveSourceIP: false
```
Comment (howardjohn) [inline manifests/charts/global.yaml:62]: We do not need to add a new option. Users can set it directly with `pilot.env.TPROXY_PRESERVE_SOURCE_IP=true`

File: pilot/pkg/features/pilot.go [hunk 4]
```diff
@@ -321,4 +321,11 @@ var (
 		"If this is set to true, one Istiod will control remote clusters including CA.").Get()
 	EnableCAServer = env.RegisterBoolVar("ENABLE_CA_SERVER", true,
 		"If this is set to false, will not create CA server in istiod.").Get()
+
+	EnableTProxyPreserveSourceIP = env.RegisterBoolVar(
```
Comment (rlenglet) [inline pilot/pkg/features/pilot.go:325]: As I commented before, please remove this option.
Unconditionally use the original_src filter when in the TPROXY mode. Don't make this optional.
I don't see any downside to doing this.

**Review by rlenglet — CHANGES_REQUESTED** [review]

File: manifests/charts/global.yaml (hunk 3 above)
Comment (gmemcc) [reply manifests/charts/global.yaml]: OK, this commit reverted.

File: pilot/pkg/features/pilot.go (hunk 4 above)
Comment (gmemcc) [reply pilot/pkg/features/pilot.go]: Option has been removed, thanks!

**Review by rlenglet — APPROVED** [review]
Awesome, thank you!

**Comment by mandarjog** [comment]
We need God mode lgtm from @howardjohn 
We should get this in.

/cherry-pick release-1.6

**Review by howardjohn — APPROVED** [review]
we really need need some tests for this feature if we want to be serious about it

**Comment by satishmane** [comment]
@howardjohn 

Added sidecar.istio.io/interceptionMode: TPROXY to pod, it is crashing continuously.

my istio version 1.6.7
kubernetes version : GKE 1.14

**Comment by rlenglet** [comment]
@satishmane please file a GitHub issue with more details.

**Pull request merged**